File: dw/p3/p3.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class GreedMethod():
    # ファイルからテストデータ読み込み
    def parse(self, filename):
        data = dict()
        with open(filename) as f:
            for line in f:
                data_set = line.strip('\n').split(' ')
                data[data_set[0]] = data_set[1].split(',')
        return data

    def __init__(self,filename):
        test_data =  self.parse(filename)
        s = map(lambda x: int(x), test_data.get('s'))
        t = map(lambda x: int(x), test_data.get('t'))
        self.all_st = range(min(s + t),max(s+t))
        self.t_s = sorted(map(lambda x: range(x[0], x[1]), zip(s, t)), key=lambda x:len(x))

    def exclude_list(self,sublist, all_list):
        subset = set(sublist)
        allset = set(all_list)
        if subset.issubset(allset):
            return list(allset.difference(subset))
        else:
            return None

    def exclude_list_list(self,sublist, all_list):
        return filter(lambda x: x != sublist, all_list)


    def calc(self,sublist, alllist, count):
        for sub in sublist:
            if set(sub).issubset(set(alllist)):
                return self.calc(self.exclude_list_list(sub, sublist), self.exclude_list(sub, alllist), count+1)
            else:
                continue
        return count

File: dw/p3/p3.clj
(require '[clojure.set])

(def n 5)
(def s '(1 2 4 6 8))
(def t '(3 5 7 9 10))

(defn zip [a b]
  (for [i (range (count a))]
     (list (nth a i) (nth b i))))

(def set-all (set '(1 2 3 4 5 6 7 8 9)))
(def list-subset (list (set '(1 2)) (set '(2 3 4)) (set '(4 5 6)) (set '(6 7 8)) (set '(8 9))))

(println (zip [1 2 3 4] [2 3 4 5]))

(defn min-numlist [l]
  (first (sort l)))

(defn min-count [list-subset]
  (min-numlist (map #(count %) list-subset)))

(defn search [list-subset set-all count]
  (let [subset-list (filter #(= (count %) (min-count list-subset)) list-subset)]
    (if (not (empty? subset-list))
      (let [subset (nth subset-list 0)]
        (if (clojure.set/subset? subset set-all)
          (search (clojure.set/difference set-all (rest subset-list)) (inc count))
          count)
        count))))
;
;(def set-p (set '(1 2 3 4 5 6 7 8 9)))
;(def set-st (list (set '(1 2)) (set '(2 3 4)) (set '(4 5 6)) (set '(6 7 8)) (set '(8 9))))
;
;(println (search list-subset set-all 0))

;(println (clojure.set/subset? (nth list-subset 0) set-all))
;(println (nth list-subset 0) )
;(println (clojure.set/difference set-all (nth list-subset 0)))
